keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -795,79 +846,192 @@ enhancement:
 * **Are there any tests for feature enablement/disablement?**
   The e2e framework does not currently support enabling and disabling feature
   gates. However, unit tests in each component dealing with managing data created
-  with and without the feature are necessary. At the very least, think about
-  conversion tests if API types are being modified.
+  with and without the feature are necessary and will be added before
+  before the transition to beta.
 
 ### Rollout, Upgrade and Rollback Planning
 
-Will be added before the transition to beta.
-
 * **How can a rollout fail? Can it impact already running workloads?**
 
+A rollout happens in at least two phases:
+1. Updating the cluster so that the `CSIStorageCapacity` API is enabled in the apiserver
+   and the kube-scheduler uses that information *for drivers which have opted into this*.
+2. CSI driver installations get updated such that they produce `CSIStorageCapacity` objects
+   and enable usage of those objects in their `CSIDriver` object.
+
+In the first phase, scheduling of pods should continue as before
+because no CSI driver has opted into the feature yet. If it doesn't
+continue, then the implementation is faulty and the feature needs to
+be disabled again until a fix is available. Then second phase gets
+skipped and the cluster operates as before.
+
+If the second phase fails because a driver malfunctions or overloads
+the apiserver, then it can be rolled back and scheduling again happens
+without using storage capacity information.
+
+In none of these cases are running workloads affected.
+
 * **What specific metrics should inform a rollback?**
 
+One is an increased number of pods that are not getting scheduled with
+events that quote `node(s) did not have enough free storage` as reason
+when the cluster is not really running out of storage capacity.
+
+Another is a degradation in apiserver metrics (increased CPU or memory
+consumption, increased latency).
+
 * **Were upgrade and rollback tested? Was upgrade->downgrade->upgrade path tested?**
 
+Not yet, but will be done manually before transition to beta.
+
 * **Is the rollout accompanied by any deprecations and/or removals of features,
   APIs, fields of API types, flags, etc.?**
 
-### Monitoring requirements
+No.
 
-Will be added before the transition to beta.
+### Monitoring requirements
 
 * **How can an operator determine if the feature is in use by workloads?**
 
+The feature is used if all of the following points apply:
+- `CSIDriver.spec.storageCapacity` is true for a CSI driver.
+- There are storage classes for that driver.
+- Volumes are using those storage classes.
```

Comment:
This is a fairly convoluted way to detecting that.
Can we expose some metrics for this purpose?

Reply:
That would have to go into kube-scheduler. What should that metric count? Number of times that a storage capacity was the deciding factor between using or not using a node?

There are no metrics at all that expose why nodes are chosen as far as I know - or at least there are none in the part that is involved with volumes.

IMHO this needs more thought; a new "enhance metrics for volume scheduling" KEP might be a better place for it.

Reply:
OK - let me take a step back here.

Basically - this isn't a feature of a "running workload" per-se. It only affects on which node (if any) it will get scheduled.

So let's maybe add something like that. at the beginning.

As for the second point, I think the scheduler metric is exactly what we want.
I don't care that much if it was a deciding factor or not. But I would like to see if we were attempting to use to that information.

As a first step in this KEP, I was thinking about a metric like: "I (scheduler) wanted to check CSIStorageCapacity object and it didn't exist for that Node". But now I realize that it doesn't really have to be an error, because there may not be particular type of storage on that node, right?

So let's maybe make one more step back and think about producing CSIStorageCapacity object. In L662 you propose a central-provisioner that will be GC-ing orphaned objects. Can we extend it a bit so that:
- it will be exposing a metric for number of desired CSIStorageCapacity object splitted by type
- ?

Reply:
> Basically - this isn't a feature of a "running workload" per-se. It only affects on which node (if any) it will get scheduled.
> 
> So let's maybe add something like that. at the beginning.

In this section, right?

> As for the second point, I think the scheduler metric is exactly what we want.
> I don't care that much if it was a deciding factor or not. But I would like to see if we were attempting to use to that information.

I've asked on #sig-scheduling about it: https://kubernetes.slack.com/archives/C09TP78DV/p1612270660035300

> As a first step in this KEP, I was thinking about a metric like: "I (scheduler) wanted to check CSIStorageCapacity object and it didn't exist for that Node". But now I realize that it doesn't really have to be an error, because there may not be particular type of storage on that node, right?

Correct. I think it is fine to count this, admins just need to be aware that it may be normal. Is that then still worthwhile?

We can also count other things:
- number of times the volume binder was invoked (= total number of scheduling decisions)
- number of times the volume binder couldn't let a pod proceed: should be less than the total count and maybe gives an impression whether the storage subsystem is keeping up
- number of times that a pod could not proceed because storage was exhausted: again, should be a smaller number

> So let's maybe make one more step back and think about producing CSIStorageCapacity object. In L662 you propose a central-provisioner that will be GC-ing orphaned objects. Can we extend it a bit so that:
> 
>     * it will be exposing a metric for number of desired CSIStorageCapacity object splitted by type

Which type?

We can distinguish between objects for different storage classes and use the storage class name as label, but for topology there is no good way to synthesize a string that could be used as value for a "topology" label.

I think we should focus on "desired number of objects" and "actual number of objects" (may be higher while some became obsolete!) regardless of type.

I can also add the workqueue length.

Reply:
> Correct. I think it is fine to count this, admins just need to be aware that it may be normal. Is that then still worthwhile?

let's ignore this for now

> We can distinguish between objects for different storage classes and use the storage class name as label, but for topology there is no good way to synthesize a string that could be used as value for a "topology" label.

I meant by 'storage class".

> I think we should focus on "desired number of objects" and "actual number of objects" (may be higher while some became obsolete!) regardless of type.

Yes - that's exactly what I had on my mind.
The actual we can take by listing objects, but if you want to expose that too, it would be even better.

Can you please add this?

Reply:
Done. I left out more advanced metrices for scheduling because this needs more overall planning.